### DISCUSSION TITLE
GPT-3.5 vs GPT-4 doing music programming with Glicol

### DISCUSSION THREAD

**Discussion by chaosprint** [chaosprint/glicol#125] [Show and tell]
https://user-images.githubusercontent.com/35621141/225110453-97d03f19-7969-49c7-a8ea-d880de0c4308.mp4

What you see is the result of GPT generation after I send the model a tutorial on Glicol, focusing on the basic structure, node usages of `sin`, `mul`, `add`, `seq`, `sp`, `mix` and `plate`.

GPT-3.5 reacted quite poorly on the generation, and I have to correct it over and over again:
> main mistake: don't know how to do `fm` and use non-existed `echo` and `whoosh` (where do you hear them from?)

```
// the first response of GPT-3.5
~bd: seq 60 _60 _60 _60 >> sp \808bd >> mul 0.8
~sd: seq _ 60 __60 _60 >> sp \snare2 >> mul 0.5
~oh: seq _ 60 _60 _60 _60 _60 _60 _60 60 >> sp \openhh >> mul 0.3
~ch: seq _ _ _ 60 >> sp \808ch >> mul 0.2
~clap: seq 60 _ 60 __60 >> sp \808clap >> mul 0.5

~mod1: sin 0.1 >> mul 0.3 >> add 0.5
~mod2: sin 0.15 >> mul 0.2 >> add 0.8
~mod3: sin 0.05 >> mul 0.1 >> add 0.2

~fm: sin ~mod1 >> mul ~mod2 >> add ~mod3 >> mul 450.0 >> add 550.0

~bass: seq _ 60 _ 60 >> sp \bass3 >> mul 0.6
~lead: seq _ 67 67 _67 _ >> sp \pluck >> mul 0.5

out: mix ~bd ~sd ~oh ~ch ~clap ~lead ~bass >> echo 0.2 >> whoosh 0.1 >> plate 0.3

```

While what you see on the GPT-4 was the first version of the generation, and I only need to fix the mistakenly use of `sin`:

```
// first time GPT-4 response
~kick: seq 60 _ _ _ >> sp \808bd >> mul 0.6
~snare: seq _ _ 60 _ >> sp \808sd >> mul 0.6
~hihat: seq 60 60 60 60 >> sp \808ch >> mul 0.3
~melody: seq 60 62 64 65 67 65 64 62 >> sin >> mul 0.4
~bass: seq 48 48 _ 48 50 50 _ 50 >> sp \bass3 >> mul 0.5

out: mix ~k.. ~s.. ~h.. ~melody ~bass >> plate 0.1
```

My opinion: GPT-4 is significantly smarter:
- pick up the syntax super in `mix` node, although it's still not getting the central idea
- know how to write concise `seq`

---
After two iterations of dialogues, the model writes this. I did not modify anything.
You can play it at https://glicol.org/
```
~kick: speed 4.0 >> seq 60 _ _ _ >> sp \808bd >> mul 0.6
~snare: speed 4.0 >> seq _ _ 60 _ >> sp \808sd >> mul 0.6
~hihat: speed 16.0 >> seq 60 62 60 62 >> sp \808ch >> mul 0.3
~melody: speed 4.0 >> seq 60 67 64 62 65 69 62 60 >> sp \blip >> mul 0.4
~bass: speed 4.0 >> seq 33 31 36 38 40 43 38 36 >> sawsynth 0.01 0.1 >> mul 0.5 >> lpf 800.0 1.0

out: mix ~k.. ~s.. ~h.. ~melody ~bass >> plate 0.1
```

The only issue is the speed, which is not difficult to ask for a fix:
```
~kick: speed 1.0 >> seq 60 _ _ _ >> sp \808bd >> mul 0.6
~snare: speed 1.0 >> seq _ _ 60 _ >> sp \808sd >> mul 0.6
~hihat: speed 4.0 >> seq 60 62 60 62 >> sp \808ch >> mul 0.3
~melody: speed 1.0 >> seq 60 67 64 62 65 69 62 60 >> sp \blip >> mul 0.4
~bass: speed 1.0 >> seq 33 31 36 38 40 43 38 36 >> sawsynth 0.01 0.1 >> mul 0.5 >> lpf 800.0 1.0

out: mix ~k.. ~s.. ~h.. ~melody ~bass >> plate 0.1
```

**Comment by nikivdev** [comment]
> What you see is the result of GPT generation after I send the model a tutorial on Glicol

Can you share the actual prompt? I am bit confused by what you mean by this.

**Reply by chaosprint** [reply]
<details><summary>I used Chinese to make the prompt. I guess it won't make lots of difference for the model but you may need to use a translator to understand it. It was just a quick test, but the difference was very clear. I can try to write an English prompt in the future. </summary>
现在你是一个音乐编程语言的学习者，我来教你一门新语言叫 Glicol。这门语言主要用于 live coding 即兴演出。在这门语言中，我们要做的就是不断创建 node，并连接这些 node 来构造一个音频 graph。以下是一个简单的例子：
`o: sin 440 >> mul 0.1`
通过这个例子，我们学到了 `sin` 和 `mul` 两个节点，后面可以跟一个单独的参数，`sin`的参数表示 sine wave oscillator 的频率，而 `mul` 的参数表示比例。每个node的输入和输出都是音频float数字流，而连接node则使用两个大于号`>>`。
可以把一个chain的node起一个reference名字，这在更复杂的代码中非常有用。

看下面这个例子。`sin` 的输出信号的数字范围在-1到1之间，通过mul 和 add就可以扩展到0.2到0.8之间。给这个chain的node起一个reference叫`~mod1`。而在第二行的代码中，mul的参数就是这个`～mod`信号，也就会在0.2到0.8中规律地变化。这种效果就是 amplitude modulation。第二行的代码reference是`o`，只有不以tilde `~`开头的chain会被播放。这点一定要注意。
```
~mod1: sin 0.1 >> mul 0.3 >> add 0.5
o: sin 440 >> mul ~mod1
```
再来看个frequency modulation的例子，还是用这些node：
```
out: sin ~fm >> mul 0.1
~fm: sin 200 >> mul 450.0 >> add 550.0
```
这会播放一个很像钟声的音色。
一个比较复杂的node是 seq，通常和 sp 这个节点配合使用：

```
o: seq 60 _ 60__60 _60 >> sp \bd
```
seq 可以接受任意数量的数字或者下划线作为参数。其中数字表示MIDI音符，下划线表示休止。这些音符和休止符一共占据一个小节的长度。由于系统的默认bpm是120，signature是4/4，那么一个小节的默认长度就是两秒。这一个小节的音符时值，首先按照空格来均分。在上面的例子中，一个小节的时间就被先均分成`60`, `_60`, `60__60`, `_60`这四个部分。 `60__60`, `_60`这两个中间没有空格的数字和下划线组合叫做compound note。对于这些compound note，可以按照数字和下划线的总数量再均分。因此，整个小节的第三拍中，会先播放60，等待两个十六分音符后，再播放下一个60。在整个小节的第四拍中，会先休止一个八分音符，然后再播放一个八分音符的60，也就是俗称的“反拍”。

sp 节点会根据输入的信息来播放采样，60就按原音调原速度播放，72则加速一倍，音调也会对应MIDI一样高八度。


另外一个重要的节点是mix，看下面这个例子：

```
~t1: seq 60 _60 _67 _65 >> sp \bass3 >> mul 0.3
~t2: seq 60 60 60 60 >> sp \808bd >> mul 0.6
out: mix ~t.. >> plate 0.1
```

这个例子里，mix 接受一个reference 作为参数，但是这个reference用两个英文句号跟在后面作为语法糖表示模糊匹配。所以在这个例子里等效于 `mix ~t1 ~t2`。如果有~t3, ~t4也会被mix进来。最后给mix后的信号加上一个混响0.1。`plate 0.1`能达到最好的效果，以后都这样用。

现在我告诉你可用的采样：
```
['808bd', '808ch', '808ht', '808clap', '808lt', '808oh', '808sd', 'blip', 'bin', 'bass3', 'boip', 'bpblart', 'bong', 'can', 'casiohigh', 'casionoise', 'casiolow', 'closedhh', 'clubkick', 'crash', 'cb', 'fat808sub', 'glass', 'giveit', 'hit1', 'echo', 'guitar', 'industrial', 'hit3', 'hit2', 'kick1', 'kick2', 'latibro', 'junglesine', 'perc1', 'openhh', 'snare1', 'pad', 'rm', 'sid', 'pluck', 'snare2', 'stab', 'ride', 'talk1', 'perc2', 'talk2', 'moog', 'tok', 'sax', 'tink', 'wind', 'whoosh', 'ufo']
```

你来作个曲子吧，如果有不会的可以问，让我来补充。
</details>

**Reply by chaosprint** [reply]
<details><summary>The English translation by GPT-4</summary>
Now you are a learner of a music programming language, and I will teach you a new language called Glicol. This language is mainly used for live coding improvisational performances. In this language, what we need to do is to continuously create nodes and connect these nodes to construct an audio graph. Here is a simple example:
o: sin 440 >> mul 0.1
Through this example, we learned two nodes, sin and mul, which can be followed by a single parameter. The parameter of sin represents the frequency of the sine wave oscillator, and the parameter of mul represents the ratio. The input and output of each node are audio float number streams, and nodes are connected using two greater-than signs >>.
You can give a chain of nodes a reference name, which is very useful in more complex code.

Look at the example below. The output signal of sin has a digital range between -1 and 1, and it can be expanded to a range between 0.2 and 0.8 through mul and add. Give this chain of nodes a reference called ~mod1. In the second line of code, the parameter of mul is the ~mod signal, which will vary regularly between 0.2 and 0.8. This effect is called amplitude modulation. The reference of the second line of code is o, and only chains that do not start with a tilde ~ will be played. Be sure to pay attention to this point.

```
~mod1: sin 0.1 >> mul 0.3 >> add 0.5
o: sin 440 >> mul ~mod1
```

Let's look at an example of frequency modulation using these nodes:

```
out: sin ~fm >> mul 0.1
~fm: sin 200 >> mul 450.0 >> add 550.0
```
This will play a sound very similar to the chime of a clock.
A more complex node is seq, which is often used in combination with the sp node:

```
o: seq 60 _ 60__60 _60 >> sp \bd
```

seq can accept any number of numbers or underscores as parameters. The numbers represent MIDI notes, and the underscores represent rests. These notes and rests together occupy the length of one measure. Since the system's default bpm is 120 and the signature is 4/4, the default length of a measure is two seconds. The note values of this measure are first evenly divided by spaces. In the example above, the time of a measure is first evenly divided into four parts: 60, _60, 60__60, and _60. The combinations of numbers and underscores without spaces in the middle, such as 60__60 and _60, are called compound notes. For these compound notes, they can be further evenly divided according to the total number of numbers and underscores. Therefore, in the third beat of the entire measure, the 60 will be played first, and after waiting for two sixteenth notes, the next 60 will be played. In the fourth beat of the entire measure, there will be a rest for an eighth note, and then a 60 will be played as an eighth note, which is commonly known as an "offbeat."

The sp node will play samples according to the input information. 60 will play at the original pitch and speed, while 72 will play at twice the speed and an octave higher in pitch, just like MIDI.

Another important node is mix. Look at the following example:

```
~t1: seq 60 _60 _67 _65 >> sp \bass3 >> mul 0.3
~t2: seq 60 60 60 60 >> sp \808bd >> mul 0.6
out: mix ~t.. >> plate 0.1
```

In this example, mix accepts a reference as a parameter, but the reference is followed by two periods as syntactic sugar to represent a fuzzy match. So in this example, it is equivalent to mix ~t1 ~t2. If there are ~t3 and ~t4, they will also be mixed in. Finally, add a reverb of 0.1 to the mixed signal. The plate 0.1 achieves the best effect, so use it this way in the future.

Now I'll tell you the available samples:

```
['808bd', '808ch', '808ht', '808clap', '808lt', '808oh', '808sd', 'blip', 'bin', 'bass3', 'boip', 'bpblart', 'bong', 'can', 'casiohigh', 'casionoise', 'casiolow', 'closedhh', 'clubkick', 'crash', 'cb', 'fat808sub', 'glass', 'giveit', 'hit1', 'echo', 'guitar', 'industrial', 'hit3', 'hit2', 'kick1', 'kick2', 'latibro', 'junglesine', 'perc1', 'openhh', 'snare1', 'pad', 'rm', 'sid', 'pluck', 'snare2', 'stab', 'ride', 'talk1', 'perc2', 'talk2', 'moog', 'tok', 'sax', 'tink', 'wind', 'whoosh', 'ufo']
```

Now it's your turn to create a piece of music. If there's anything you don't understand, feel free to ask, and I'll provide further explanations.

</details>
